### PR TITLE
router: Remove PathMatchType and PathMatchCriterion

### PR DESCRIPTION
**Pull request by tpetkov-VMW**
Signed-off-by: Toma Petkov <tpetkov@vmware.com>

Commit Message: Remove PathMatchType enum and PathMatchCriterion class, as they aren't used anywhere.
Additional Description: PathMatchCriterion class is never instantiated anywhere, besides for inheritance and testing.
Risk Level: Low
Testing: Unit test (test/common/router)
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by tpetkov-VMW** [comment]
Proposal to remove ``PathMatchCriterion``, based only on the codebase. 
If there are, for example, pending changes that depend on it, I'll just close the PR.

**Review by mattklein123 — APPROVED** [review]
I don't recall the history on this. It seems fine to remove. Thank you!

**Comment by htuch** [comment]
+1

**Comment by stevenzzzz** [comment]
hi there, this PR breaks our system, I think we need to reverse this PR if no strong opinion. 

TLDR we need this field to take a peek into how route selection works. 

Please see the original issue/PR for reference: https://github.com/envoyproxy/envoy/pull/2531

**Comment by stevenzzzz** [comment]
@htuch

**Comment by mattklein123** [comment]
Sure feel free to revert but please add comments/tests that make it clear this is used in private extensions so it doesn't get removed again.

**Comment by stevenzzzz** [comment]
thanks. @adisuissa could you please revert this as we've discussed.
